Add AMI build job to build workflow

jsf9k opened the pull request:
## 🗣 Description ##

This pull request adds a `build` task to the `build` workflow.  This new task goes through the process of building a new AMI but does not actually register it.

This pull request is blocked until we move to Packer 1.7 (see #70 and cisagov/setup-env-github-action#32), since [that is where support for the `skip_create_ami` key was added for Packer configurations](https://github.com/hashicorp/packer/blob/master/CHANGELOG.md#improvements-3) (see the first bullet).

## 💭 Motivation and context ##

This is a useful feature to add since it will catch changes that cause the AMI build process to fail.  Currently these changes are only caught when we create a pre-release or release.

N.B.:  Once we have dev, staging, and production COOL environments we could allow this job to go ahead and register the AMI in the dev environment.  That would allow us to fully test in the dev environment before we create a pre-release (for staging) and/or a release (for production).

## 🧪 Testing ##

cisagov/setup-env-github-action#32 was merged and now all `pre-commit` hooks and GitHub Actions workflows pass. 

## ✅ Checklist ##

* [x] This PR has an informative and human-readable title.
* [x] Changes are limited to a single goal - _eschew scope creep!_
* [x] All relevant type-of-change labels have been added.
* [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
* [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
* [x] Tests have been added and/or modified to cover the changes in this PR.
* [x] All new and existing tests pass.
* [x] Add `build/build` job to the list of GH Actions jobs that must pass before a PR can be merged.

jsf9k committed:
Add a test AMI build to the build workflow

The test build goes through the process of building an AMI but does
not actually create it.  This is an good thing to add since it will
catch changes that cause the AMI build process to fail.  Currently
these changes are only caught when we create a pre-release or release.

jsf9k committed:
Do not run the build job unless the other jobs have passed

The build job can be an expensive test, so we don't want to bother
running it unless the other jobs pass.

felddy reviewed (approved):
This is a good improvement.  It should catch a whole new class of errors.
Strong work. 💪

dav3r commented on .github/workflows/build.yml:
```diff
@@ -133,3 +136,95 @@ jobs:
+          python -m pip install --upgrade pip
+          pip install --upgrade \
+                      --requirement requirements.txt
+      - name: Install ansible roles
```
```suggestion
      - name: Install Ansible roles
```

mcdonnnj commented on .github/workflows/build.yml:
```diff
@@ -133,3 +136,95 @@ jobs:
+            "https://releases.hashicorp.com/terraform/${TERRAFORM_VERSION}/${TERRAFORM_ZIP}"
+          sudo unzip -d /opt/terraform \
+            ${{ env.CURL_CACHE_DIR }}/"${TERRAFORM_ZIP}"
+          sudo ln -s /opt/terraform/terraform /usr/bin/terraform
```
Mirroring the decision in https://github.com/cisagov/skeleton-generic/pull/82#discussion_r667214249

```suggestion
```

mcdonnnj commented on .github/workflows/build.yml:
```diff
@@ -133,3 +136,95 @@ jobs:
+            --time-cond ${{ env.CURL_CACHE_DIR }}/"${PACKER_ZIP}" \
+            --location \
+            "https://releases.hashicorp.com/packer/${PACKER_VERSION}/${PACKER_ZIP}"
+          sudo unzip -o -d /usr/local/bin \
+            ${{ env.CURL_CACHE_DIR }}/"${PACKER_ZIP}"
```
Mirroring https://github.com/cisagov/skeleton-generic/pull/82#discussion_r667207173

```suggestion
          sudo unzip -d /opt/packer \
            ${{ env.CURL_CACHE_DIR }}/"${PACKER_ZIP}"
          sudo mv /usr/local/bin/packer /usr/local/bin/packer-default
          sudo ln -s /opt/packer/packer /usr/local/bin/packer
```

mcdonnnj commented on .github/workflows/build.yml:
```diff
@@ -133,3 +136,95 @@ jobs:
+          python -m pip install --upgrade pip
+          pip install --upgrade \
+                      --requirement requirements.txt
+      - name: Install ansible roles
```
```suggestion
      - name: Install Ansible roles
```

jsf9k replied:
See commit e95501d.

dav3r reviewed (approved):
Excellent! 🎸 
I only noted one small thang.

mcdonnnj commented on .github/workflows/build.yml:
```diff
@@ -133,3 +136,95 @@ jobs:
+            ${{ hashFiles('**/requirements.txt') }}"
+          restore-keys: |
+            ${{ env.BASE_CACHE_KEY }}
+      - name: Install Packer
```
Mirroring https://github.com/cisagov/skeleton-generic/pull/82/files#r667205518

```suggestion
      - name: Setup curl cache
        run: mkdir -p ${{ env.CURL_CACHE_DIR }}
      - name: Install Packer
```

mcdonnnj commented on .github/workflows/build.yml:
```diff
@@ -133,3 +136,95 @@ jobs:
+            ${{ env.BASE_CACHE_KEY }}
+      - name: Install Packer
+        run: |
+          mkdir -p ${{ env.CURL_CACHE_DIR }}
```
```suggestion
```

mcdonnnj reviewed (changes requested):
This is a great improvement. I have some suggestions that are mostly to bring this into alignment with the work in https://github.com/cisagov/skeleton-generic/pull/82.

jsf9k committed:
Capitalize Ansible

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

jsf9k committed:
Break curl cache directory creation into its own workflow step

This mirrors what was done in cisagov/skeleton-generic#82.

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Remove needless linking of Terraform executable

This mirrors what was done in cisagov/skeleton-generic#82.

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Make the Packer install non-destructive

This mirrors what was done in cisagov/skeleton-generic#82.

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

mcdonnnj commented on .github/workflows/build.yml:
```diff
@@ -133,3 +136,97 @@ jobs:
+  build:
+    # The AMI build process is an expensive test (in terms of time) so
+    # let's not run it unless the other jobs succeed.
+    needs:
+      - lint
+      - test
```
This is how it's done in [the cisagov/skeleton-python-library build job](https://github.com/cisagov/skeleton-python-library/blob/4ef2364701c0db074c423f740062b2d7c197bea8/.github/workflows/build.yml#L113). I'm fine if we want to switch to this style, but I'd like to use the same format for similar functionality whichever way we decide.

```suggestion
    needs: [lint, test]
```

mcdonnnj commented on .github/workflows/build.yml:
```diff
@@ -133,3 +136,97 @@ jobs:
+      # actions/setup-python step above.  This can hose other tasks
+      # that are expecting to find the system Python at that location,
+      # though, so we undo this change after running Packer.
+      - name: Create a /usr/bin/python3 symlink to the installed python
```
```suggestion
      - name: Create a /usr/bin/python3 symlink to the installed Python
```

mcdonnnj commented on .github/workflows/build.yml:
```diff
@@ -133,3 +136,97 @@ jobs:
+        run: |
+          packer build --timestamp-ui -var skip_create_ami=true \
+          src/packer.json
+      - name: Remove /usr/bin/python3 symlink to the installed python
```
```suggestion
      - name: Remove /usr/bin/python3 symlink to the installed Python
```

mcdonnnj reviewed (approved):
just some last suggestions. Thanks again for thinking of and implementing this improvement!

jsf9k committed:
Capitalize Python for consistency

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Change list style for consistency

This is how it's done in [the cisagov/skeleton-python-library build job](https://github.com/cisagov/skeleton-python-library/blob/4ef2364701c0db074c423f740062b2d7c197bea8/.github/workflows/build.yml#L113).

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

mcdonnnj committed:
Merge branch 'develop' into improvement/add-ami-build-to-build-workflow
